Refactor Focus page (#10749)
flutter/website: src/content/ui/interactivity/focus.md

description: How to use the focus system in your Flutter app.
---

This article explains how to direct keyboard input.
If your app uses a physical keyboard, like desktop and web apps,
keep reading. If your app won't use a physical keyboard,
you can skip this guide.

## Overview

Flutter includes a focus system that directs keyboard input
to a particular part of an app.
To "focus" the input to part of an app, app users can take action
depending on their input method.

* To change focus on touch interfaces,
  app users tap or click on the desired UI element.

* To change focus on traditional interfaces,
  app users can use the mouse to click on the desired UI element
  or press a keyboard shortcut to move focus.

Once in focus, text entered with the keyboard flows to that element.
Text input could control objects or display text in a text field.
This continues until the focus moves to another element in the app.

> :::secondary What you'll learn

Reviewer: Why did you replace the list of focus use cases with "what you'll learn"? This isn't a tutorial. I put those use cases there for a VERY SPECIFIC reason. I wanted to make it very clear what situations might require fiddling with the focus. The "what you'll learn" doesn't convey the same info. And since we've already said that this guide doesn't apply to everyone, why treat it like a general tutorial?

Author: The list of use cases that Greg added in his [initial PR](https://github.com/flutter/website/pull/6243/commits/944b5f33764406ea46efbf7374150a2d8a4eba8c#diff-a7a3d5017af0d2e57bf7203482edf9adbfbcd333fcda4ce15ad4540832affd5aR29-R39) read as lacking context, so they were moved to [When to use the focus system](https://flutter-docs-prod--pr10749-fix-10168-copy-qvo61uju.web.app/ui/interactivity/focus#when-to-use-the-focus-system).

**Don't allocate a new `FocusNode` object for each build.**
: This can cause memory leaks.
  It might also cause a loss of focus when the widget rebuilds while
  the node has focus.

**Don't use the same `FocusNode` for multiple widgets.**
: If you do, the widgets fight over managing the attributes of the node.
  That might result in an unexpected outcome.

**Don't set attributes on focus nodes without using `Focus.withExternalFocusNode()`**
: If you want to include attributes on a FocusNode,
  create a FocusNode outside of the Focus widget.
  This prevents you from unintentionally overwriting any attributes
  you set on the [`Focus`][] widget.

  This applies to a `FocusScopeNode` that a [`FocusScope`][] widget
  manages.

## Tell a node to give up focus

The `FocusNode.unfocus()` API tells a node to "give up the focus".
While it does remove focus from the node,
you can't "unfocus" all nodes.
If you remove focus from a node, it must pass the focus somewhere else,
since there is _always_ a primary focus. When a node calls `unfocus()`,
focus can pass to one of two nodes.
It can pass to either the nearest `FocusScopeNode`
or a node in the same scope that had focus earlier.

> Which node Flutter chooses depends upon the `disposition` argument
> you give to `unfocus()`.
> When calling `unfocus()`, you can set the `disposition` argument.
> It allows two modes for passing focus:

Reviewer: This text is very verbose and duplicative. Why not make it much more concise, such as:

```suggestion
Flutter chooses a new focus node depending on the `disposition` argument
you pass to `unfocus()`:
```

> * [`UnfocusDisposition.scope`][]:
>   The argument defaults to this option.
>   It gives the focus to the nearest parent focus scope.
>   If then something moves the focus to the next node with `FocusNode.nextFocus`,
>   the app starts with the "first" focusable item in the scope.

Reviewer: Again, I'd like to see this more concise:

```suggestion
* [`UnfocusDisposition.scope`][] (default):
  Flutter passes focus to the nearest parent focus scope.
  After this, calling `FocusNode.nextFocus()` moves focus to the first
  focusable widget in that scope.
```

* `UnfocusDisposition.previouslyFocusedChild`:
  The `previouslyFocusedChild` disposition searches the scope to find the
  most recently focused child and requests focus on it.
  If no child had focus before, this disposition works like `scope`.

To better control where the focus passes when you remove it from a node,
you have two options.
You can choose to set focus to another node using `requestFocus()`
or use the focus traversal mechanism.
To find another node, this mechanism can use the `focusInDirection`,
`nextFocus`, or `previousFocus` methods on `FocusNode`.

## Allow a control to receive focus

The `Focus` widget owns and manages a focus node.
It performs the following functions as the workhorse of the focus system:

* Attaches and detaches the focus node it owns from the focus tree
* Manages the attributes and callbacks of the focus node
* Enables discovery of focus nodes attached to the widget

>   tree through its static functions

Reviewer: Why did you replace this text with a bulleted list? Particularly following a header, it should prose, not a list. It's much harder to parse this way.
